compiler/rustc_typeck/src/check/pat.rs

```diff
@@ -663,6 +663,46 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 ast::Mutability::Not => "",
             };
 
+            let mut_var_suggestion = 'block: {
+                if !matches!(mutbl, ast::Mutability::Mut) {
+                    break 'block None;
+                }
+
+                let ident_kind = match binding_parent {
+                    hir::Node::Param(_) => Some("parameter"),
+                    hir::Node::Local(_) => Some("variable"),
+                    hir::Node::Arm(_) => Some("binding"),
+
+                    // Provide diagnostics only if the parent pattern is struct-like,
+                    // i.e. where `mut binding` makes sense
+                    hir::Node::Pat(Pat { kind, .. }) => match kind {
+                        PatKind::Struct(..)
+                        | PatKind::TupleStruct(..)
+                        | PatKind::Or(..)
+                        | PatKind::Tuple(..)
+                        | PatKind::Slice(..) => Some("binding"),
+
+                        PatKind::Wild
+                        | PatKind::Binding(..)
+                        | PatKind::Path(..)
+                        | PatKind::Box(..)
+                        | PatKind::Ref(..)
+                        | PatKind::Lit(..)
+                        | PatKind::Range(..) => None,
+                    },
+
+                    // Don't provide suggestions in other cases
+                    _ => None,
+                };
```

Comment:
```suggestion
                let ident_kind = match binding_parent {
                    hir::Node::Param(_) => "parameter",
                    hir::Node::Local(_) => "variable",
                    hir::Node::Arm(_) => "binding",

                    // Provide diagnostics only if the parent pattern is struct-like,
                    // i.e. where `mut binding` makes sense
                    hir::Node::Pat(Pat { kind, .. }) => match kind {
                        PatKind::Struct(..)
                        | PatKind::TupleStruct(..)
                        | PatKind::Or(..)
                        | PatKind::Tuple(..)
                        | PatKind::Slice(..) => "binding",

                        PatKind::Wild
                        | PatKind::Binding(..)
                        | PatKind::Path(..)
                        | PatKind::Box(..)
                        | PatKind::Ref(..)
                        | PatKind::Lit(..)
                        | PatKind::Range(..) => break 'block None,
                    },

                    // Don't provide suggestions in other cases
                    _ => break 'block None,
                };
```

Reply:
Oh, right. I've rewritten this code a couple of times and the match was added before the block xd